util/chunk/disk.go

```diff
@@ -0,0 +1,284 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/disk"
+	"github.com/pingcap/tidb/util/stringutil"
+)
+
+const (
+	writeBufSize = 128 * 1024
+	readBufSize  = 4 * 1024
+)
+
+var bufWriterPool = sync.Pool{
+	New: func() interface{} { return bufio.NewWriterSize(nil, writeBufSize) },
+}
+
+var bufReaderPool = sync.Pool{
+	New: func() interface{} { return bufio.NewReaderSize(nil, readBufSize) },
+}
+
+var tmpDir = path.Join(os.TempDir(), "tidb-server-hashJoin")
+
+func init() {
+	_ = os.RemoveAll(tmpDir)
+	_ = os.Mkdir(tmpDir, 0755)
+}
+
+// ListInDisk represents a slice of chunks storing in temporary disk.
+type ListInDisk struct {
+	fieldTypes []*types.FieldType
+	// offsets stores the offsets in disk of all RowPtr,
+	// the offset of one RowPtr is offsets[RowPtr.ChkIdx][RowPtr.RowIdx].
+	offsets [][]int64
+	// offWrite is the current offset for writing
+	offWrite int64
+
+	disk        *os.File
+	bufWriter   *bufio.Writer
+	diskTracker *disk.Tracker // track disk usage.
+}
+
+var chunkListInDiskLabel fmt.Stringer = stringutil.StringerStr("chunk.ListInDisk")
+
+// NewListInDisk creates a new ListInDisk with field types.
+func NewListInDisk(fieldTypes []*types.FieldType) *ListInDisk {
+	l := &ListInDisk{
+		fieldTypes: fieldTypes,
+		// TODO(fengliyuan): set the quota of disk usage.
+		diskTracker: disk.NewTracker(chunkListInDiskLabel, -1),
+	}
+	return l
+}
+
+// GetDiskTracker returns the memory tracker of this List.
+func (l *ListInDisk) GetDiskTracker() *disk.Tracker {
+	return l.diskTracker
+}
+
+// Add adds a chunk to the ListInDisk. Caller must make sure the input chk
+// is not empty and not used any more and has the same field types.
+func (l *ListInDisk) Add(chk *Chunk) (err error) {
+	if chk.NumRows() == 0 {
+		panic("chunk appended to List should have at least 1 row")
+	}
+	if l.disk == nil {
+		l.disk, err = ioutil.TempFile(tmpDir, "listInDisk")
+		if err != nil {
+			return
+		}
+		l.bufWriter = bufWriterPool.Get().(*bufio.Writer)
+		l.bufWriter.Reset(l.disk)
+	}
+	chk2 := chunkInDisk{Chunk: chk, off: l.offWrite}
+	n, err := chk2.WriteTo(l.bufWriter)
+	l.offWrite += n
+	if err != nil {
+		return
+	}
+	l.offsets = append(l.offsets, chk2.getOffsetsOfRows())
+	err = l.bufWriter.Flush()
+	if err == nil {
+		l.diskTracker.Consume(n)
+	}
+	return
+}
+
+// GetRow gets a Row from the ListInDisk by RowPtr.
+func (l *ListInDisk) GetRow(ptr RowPtr) (row Row, err error) {
+	off := l.offsets[ptr.ChkIdx][ptr.RowIdx]
+	r := io.NewSectionReader(l.disk, off, l.offWrite-off)
```

Comment:
It seems that we don't need to create a `SectionReader` object each time.
How about use `SectionReader.Seek`?

Reply:
`GetRow` is designed for concurrency in `hashJoinExec` so we need to create a new `SectionReader` here.